### DISCUSSION TITLE
[off-topic] E164, ACI, PNI, ServiceId

### DISCUSSION THREAD

**Discussion by stevie553** [AsamK/signal-cli#1323] [General]
This is off-topic and not related to signal-cli in particular, but rather Signal itself.

Then why post here? Because I believe AsamK understands how Signal internally works because of all the great work in signal-cli, but please go ahead and close this if you want to.

To not bother the community or AsamK, once I create this discussion, I'll close it if I can (this is my first starting a discussion on github so I don't know how it works).

Sorry for posting here.

I'll begin with my understanding of how Signal used to work/works, in a nutshell of course, and then I'll ask my questions.

In the past (E164 times) things were simple enough for me to understand, clients have a list of E164s, they check the server to see who is registered and who is not, and then they update their database with the results. And whenever you want to send anyone something, you ask the server to send X (content) to Y (E164). Easy enough.

Then a single UUID was introduced to replace E164. Now things changed a little bit but not too much. Like in the E164 times, clients have a list of E164s, they check the server to see who is registered and who is not, now the server not only returns the registration status, but it also returns the user's UUID. And whenever you want to send anyone something, you ask the server to send X (content) to Y (UUID). Again, easy enough.

Now at this point things were simple, users had a single identity (based on E164 since the verify v2 feature flag that switches safety numbers to use UUID instead was never turned on).

Note: I don't think I understand anything beyond this point, it's all (un)educated guesswork.

I think UUID was later renamed to ACI? I'm not sure.

PNI's were introduced and users now had two identities, one for ACI (originally UUID) and one for the brand new PNI.

Building and installing Spinner on an Android emulator and accessing the db shows that I have both ACI and PNI for my contacts in the recipient table. Which made me even more confused.

And then something called ServiceId gets introduced. To my understanding, I think both ACIs and PNIs are ServiceIds. Considering that PNP is turned off since it is at the time of writing, why do I have both in my db?

I have a lot of questions but I won't bombard you with them in another wall of text, I'll start with a simple (not very simple to me) one: What are ACIs, PNIs, and ServiceIds?

Thank you :)

Edit: I couldn't close the discussion, sorry.

**Comment by AsamK** [comment]
That's already a pretty good summary. There's been a lot of renaming and other changes going on .

A ServiceId is either a ACI or PNI or in the future also some other kind of identifier.

The ACI is the account identifier and never changes for a registered account. It gets deleted on the server some time after an account is deleted.
The PNI is the phone number identifier and never changes for a given phone number. So when you change your phone number you get a new PNI but keep your ACI.

When checking a phone number on the signal server you currently get both the PNI and the ACI of the account. If an ACI is available, messages are sent to that.
In the future with PNP the server will only send the PNI. So sending message to a new contact will send it to its PNI.

Now about the intention I'm also not completely sure. Because when the new contact responds to the message he will send it from his ACI. So the PNI is only seems relevant for sending a first message to a new contact ...

**Reply by stevie553** [reply]
First of all thank you very much replying!

> The ACI is the account identifier and never changes for a registered account. It gets deleted on the server some time after an account is deleted.

So, since the ACI never changes for the lifetime of the account and will survive re-registration/account deletion for some time like you said. That and the fact that all my contacts right now have both my ACI and PNI, how would PNP even hide anything since all contacts already know everything?

> The PNI is the phone number identifier and never changes for a given phone number. So when you change your phone number you get a new PNI but keep your ACI.

Does that mean that a given E164 **will always** have a static PNI? As in, are PNIs derived from E164s or are they randomly generated like ACIs? If so, if I change my phone number (I get a new PNI), if someone claimed my old phone number (carrier recycles it as you know), does that mean that the new person gets my old PNI? What happens if I delete my account then, do PNIs get the same treatment (deleted_accounts table) as ACIs?

> In the future with PNP the server will only send the PNI. So sending message to a new contact will send it to its PNI.

>Now about the intention I'm also not completely sure. Because when the new contact responds to the message he will send it from his ACI. So the PNI is only seems relevant for sending a first message to a new contact ...

And it even gets more confusing, in the Android client, I see there are two Protocol Stores, one for ACI and one for PNI, which means (possibly) multiple started sessions for the same contact and even more confusing, the Android client seems to store contacts' identities in the ACI store regardless of whether they were actually a contact's ACI or PNI (unless you're saving your own identity, which gets saved in the correct store)

The rabbit hole continues, technically speaking there are now two safety numbers, one for ACI identity and one for PNI identity (I don't mean the current safety number migration from E164 to ACI). If, as you said, PNI is used only for the first message, I get the [PNI] identity of the recipient then I start to communicate back with my ACI, so we exchange ACI identities, would I (or you) get a safety number change warning if one of us changes phone number? (essentially new PNI and new PNI identity and safety number)?

All that and we haven't touched on how groups will work.

Ps. Pretty cool avatar, is it some sort of chemical compound?

**Reply by AsamK** [reply]
PNP won't hide anything in that case, it will only work for new contacts.

About static PNI, that's how I think this backend code works: https://github.com/signalapp/Signal-Server/blob/main/service/src/main/java/org/whispersystems/textsecuregcm/storage/PhoneNumberIdentifiers.java#L56

Theoretically there would be two safety numbers, but I guess the apps will only show safety numbers based on ACI or no safety number at all, we'll see ...

There's been a recent change in the app code, that PNIs are prefixed with `PNI:` so they can be distinguished from ACIs at the protocol level.

**Comment by mjw3428** [comment]
Given that the change described above is now in place, how does it now work? How given a PNI, do you get the ACI for a new contact?

**Reply by AsamK** [reply]
You can't directly get the ACI for a given PNI.
You need to send a message to the PNI, which will show as a message request on the receiver's app. When the receiver accepts the message request and sends back a message, it will contain a PNI signature. This signature links the ACI and the PNI.
It's handled here in signal-cli: https://github.com/AsamK/signal-cli/blob/5e17fe8414b632246bdbd6a9051e6ee98296b2a0/lib/src/main/java/org/asamk/signal/manager/helper/IncomingMessageHandler.java#L397
